engine/consensus/dkg/processor/engine_test.go

```diff
@@ -184,68 +148,65 @@ func TestProcessDKGMessage_Valid(t *testing.T) {
 			<-doneCh
 		},
 		time.Second)
-	require.Equal(t, msg, receivedMsg)
+	require.Equal(t, expectedMsg, receivedMsg)
 }
 
-// TestProcessDKGMessage checks that incoming DKG messages are discarded with an
+// TestProcessMessage checks that incoming DKG messages are discarded with an
 // error if their origin is invalid, or if there is a discrepancy between the
 // origin defined in the message, and the network identifier of the origin (as
 // provided by the network utilities).
-func TestProcessDKGMessage_InvalidOrigin(t *testing.T) {
-
-	// initialize the list of dkg participants
-	committee := unittest.IdentifierListFixture(2)
-
-	// define epoch ID
-	epochCounter := rand.Uint64()
-
-	network := new(module.Network)
-	network.On("Register", mock.Anything, mock.Anything).
-		Return(nil, nil).
-		Once()
-
-	me := new(module.Local)
-	me.On("NodeID").Return(committee[1])
-
-	msgCh := make(chan dkg.DKGMessage)
-
-	engine, err := New(
-		zerolog.Logger{},
-		network,
-		me,
-		msgCh,
-		committee,
-		epochCounter,
-	)
-	require.NoError(t, err)
-
-	// check that the DKGMessage's Orig field is not out of index
-	msg := dkg.NewDKGMessage(
-		2,
-		[]byte{},
-		epochCounter,
+func TestProcessMessage_InvalidOrigin(t *testing.T) {
+
+	// destination engine
+	engine := createTestEngine(t, committee[dest])
+
+	// check that the Message's Orig field is not out of index
+	badIndexes := []int{-1, 2}
+	for _, badIndex := range badIndexes {
+		dkgMsg := msg.NewDKGMessage(
+			badIndex,
+			msgb,
+			engine.GetEpoch(),
+			engine.GetPhase(),
+		)
+		err := engine.Process(unittest.IdentifierFixture(), dkgMsg)
+		require.Error(t, err)
+	}
+
+	// check that the Message's Orig field matches the sender's network
+	// identifier
+	dkgMsg := msg.NewDKGMessage(
+		orig,
+		msgb,
+		engine.GetEpoch(),
+		engine.GetPhase(),
 	)
-
-	err = engine.Process(unittest.IdentifierFixture(), msg)
+	err := engine.Process(unittest.IdentifierFixture(), dkgMsg)
 	require.Error(t, err)
+}
 
-	msg = dkg.NewDKGMessage(
-		-1,
-		[]byte{},
-		epochCounter,
-	)
+// TestBroadcastMessage checks that the processor correctly wraps the message
+// data in a DKGMessage (with origin and epochCounter), and that it calls the
+// dkg contract client.
+func TestBroadcastMessage(t *testing.T) {
 
-	err = engine.Process(unittest.IdentifierFixture(), msg)
-	require.Error(t, err)
+	// sender engine
+	engine := createTestEngine(t, committee[orig])
 
-	// check that the DKGMessage's Orig field matches the sender's network
-	// identifier
-	msg = dkg.NewDKGMessage(
-		0,
-		[]byte{},
-		epochCounter,
+	expectedMsg := msg.NewDKGMessage(
+		orig,
+		msgb,
+		engine.GetEpoch(),
+		engine.GetPhase(),
```

Comment:
We are using two locks (epochCounterLock, dkgCounter) to make sure the updates to them are atomic, however, I don't think that protects us from race condition. 

Now look at the two calls `GetEpoch()` and `GetPhase()` here, if Phase was updated right after `GetEpoch` was returns and before `GetPhase` is called, then we will send a DKG message with an inconsistent epoch and phase.

Instead of using two mutex locks, we could just use one lock to lock the update to either Epoch or Phase, and reads should read both Epoch and Phase together atomically, rather than two calls.

Another comment:
It's not clear why `Engine` needs to expose the `SetEpoch` function. It seems, engine only needs to read the state of `EpochAndPhase` and never write. And there must be another instance who manages its state by calling `SetEpoch`. I think we could make a `EpochAndPhase` module, and give the writer to that module that updates the epoch and give the reader interface to this processor. It would be easier to ensure the atomicity at a module level.

Reply:
Thanks @zhangchiqing 

The `GetPhase()` method is not necessary anymore with the new DKGContractClient interface, so that is one less thing to worry about.

That being said, you bring up a good point about the `SetEpoch` function. This will expose us to all sorts of race conditions with the components that use the processor engine (like the [DKG Controller](https://github.com/onflow/flow-go/blob/martin/4462-broadcast-post/module/dkg/controller/controller.go)), and there is a similar probem with the `msgCh` which relays all messages to upstream consumers of DKGMessages.

I think the fundamental problem is that the engine is reusable across epochs by design. It would be much better if we could intantiate a new engine for every epoch, but I haven't seen anything similar in the codebase, which is why I did it this way. I am not sure if engines, which implement the network Engine interface, are meant to be singletons or if it's ok to instantiate them or dispose of them on the fly. 

What are your thoughts on this?

Reply:
@arrivets We could re-use a pattern that the collection node uses for epoch-scoped processes (in that case submitting it's vote for the root cluster QC). 

Move the epoch-scoped logic to a separate component (say `DKGRunner`). This component can run the full DKG protocol exactly once for one epoch. Then let the engine (say `DKGEngine`) be responsible strictly for detecting epoch transitions, instantiating the epoch-scoped `DKGRunner`, and starting it at the appropriate time.

In the collection node, the [`epochmgr` engine](https://github.com/onflow/flow-go/blob/663eaefd51d241f6ecf73415d3e1cf3f740527be/engine/collection/epochmgr/engine.go#L52) is the outer engine that is responsible for managing epoch-scoped components and it instantiates [several components](https://github.com/onflow/flow-go/blob/663eaefd51d241f6ecf73415d3e1cf3f740527be/engine/collection/epochmgr/engine.go#L31-L36) (and cleans up the instances for the previous epoch) each time there is an epoch transition. It should be much simpler for the DKG case since we're just dealing with one epoch-scoped component.